File: bazi.js
function calculateBazi(year, month, day, hour, gender, calendarType = 'solar') {
  try {
    // 1. 根据日历类型初始化对象
    let solar;
    let lunar;
    // 检查 hour 是否有效 (0-23)
    const hasHour = (hour !== null && hour !== undefined && hour !== "");
    
    let solarForDay, solarForHour;
    let hourGz = '';
    
    if (calendarType === 'lunar') {
      // 阴历输入：直接使用农历日期
      const yearInt = parseInt(year);
      const monthInt = parseInt(month);
      const dayInt = parseInt(day);
      
      // 检查是否是闰月（如果月份大于12，则月份-12是闰月月份）
      let isLeapMonth = false;
      let actualMonth = monthInt;
      if (monthInt > 12) {
        isLeapMonth = true;
        actualMonth = monthInt - 12;
      }
      
      // 从农历创建Lunar对象
      // lunar-javascript API: Lunar.fromYmd(year, month, day, isLeapMonth)
      try {
        if (hasHour) {
          const hourInt = parseInt(hour);
          // 处理晚子时 (23:00-23:59): 需要特殊处理
          if (hourInt === 23) {
            // 晚子时：使用下一天的农历日期
            lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt, isLeapMonth);
            const nextLunar = lunar.next(1);
            solarForDay = nextLunar.getSolar();
            solarForHour = nextLunar.getSolar();
          } else {
            // 早子时 (0) 和其他时辰：使用当前农历日期
            lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt, isLeapMonth);
            solarForDay = lunar.getSolar();
            // For hour calculation, we need to use the solar date with the hour
            solarForHour = Solar.fromYmdHms(solarForDay.getYear(), solarForDay.getMonth(), solarForDay.getDay(), hourInt, 0, 0);
          }
          solar = solarForDay;
        } else {
          lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt, isLeapMonth);
          solarForDay = lunar.getSolar();
          solarForHour = solarForDay;
          solar = solarForDay;
        }
      } catch (e) {
        // Fallback: try without leap month parameter if API doesn't support it
        try {
          if (hasHour) {
            const hourInt = parseInt(hour);
            if (hourInt === 23) {
              lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt);
              if (isLeapMonth && lunar.setLeapMonth) {
                lunar.setLeapMonth(true);
              }
              const nextLunar = lunar.next(1);
              solarForDay = nextLunar.getSolar();
              solarForHour = nextLunar.getSolar();
            } else {
              lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt);
              if (isLeapMonth && lunar.setLeapMonth) {
                lunar.setLeapMonth(true);
              }
              solarForDay = lunar.getSolar();
              solarForHour = Solar.fromYmdHms(solarForDay.getYear(), solarForDay.getMonth(), solarForDay.getDay(), hourInt, 0, 0);
            }
            solar = solarForDay;
          } else {
            lunar = Lunar.fromYmd(yearInt, actualMonth, dayInt);
            if (isLeapMonth && lunar.setLeapMonth) {
              lunar.setLeapMonth(true);
            }
            solarForDay = lunar.getSolar();
            solarForHour = solarForDay;
            solar = solarForDay;
          }
        } catch (e2) {
          throw new Error(`Invalid lunar date: ${yearInt}-${actualMonth}${isLeapMonth ? '(闰)' : ''}-${dayInt}`);
        }
      }
    } else {
      // 阳历输入：原有逻辑
      if (hasHour) {
        const hourInt = parseInt(hour);
        // 处理晚子时 (23:00-23:59): 需要特殊处理
        if (hourInt === 23) {
          // 晚子时：使用下一天的日期来计算农历日，小时干支用下一天的早子时
          const nextDay = new Date(parseInt(year), parseInt(month) - 1, parseInt(day));
          nextDay.setDate(nextDay.getDate() + 1);
          solarForDay = Solar.fromYmdHms(nextDay.getFullYear(), nextDay.getMonth() + 1, nextDay.getDate(), 0, 0, 0);
          // 小时干支用下一天的早子时 (hour 0 of next day)
          solarForHour = Solar.fromYmdHms(nextDay.getFullYear(), nextDay.getMonth() + 1, nextDay.getDate(), 0, 0, 0);
        } else {
          // 早子时 (0) 和其他时辰：使用当前日期
          solarForDay = Solar.fromYmdHms(parseInt(year), parseInt(month), parseInt(day), hourInt, 0, 0);
          solarForHour = solarForDay;
        }
        solar = solarForDay;
      } else {
        solar = Solar.fromYmd(parseInt(year), parseInt(month), parseInt(day));
        solarForDay = solar;
        solarForHour = solar;
      }
    }

    // 2. 获取农历和八字基础对象
    if (!lunar) {
      lunar = solarForDay.getLunar();
    }
    const eightChar = lunar.getEightChar();
    
    // 获取小时干支（对于晚子时，使用下一天的早子时的小时干支）
    const hourLunar = solarForHour.getLunar();
    const hourEightChar = hourLunar.getEightChar();
    hourGz = hourEightChar.getTime();
    
    let result = '';
    
    // 3. 获取当前选定时间的干支
    const yearGz = eightChar.getYear();
    const monthGz = eightChar.getMonth();
    const dayGz = eightChar.getDay();
    
    result += `${year} ${yearGz}年 ${monthGz}月 ${dayGz}日`;
    
    if (hasHour) {
      result += ` ${hourGz}时\n`;
    } else {
      result += '\n';
    }

    // 4. 计算大运 (Da Yun) - from birth to age 80
    // gender 为 'male' 传 1, 否则传 0
    const isMale = (gender === 'male' || gender === 'm' || gender === '1');
    const yun = eightChar.getYun(isMale ? 1 : 0);
    const dayunList = yun.getDaYun();
    
    // Calculate age 80 year (birth year + 80)
    const birthYear = parseInt(year);
    const age80Year = birthYear + 80;

    // 5. 过滤大运：显示从出生到80岁期间的DaYun，且必须有干支
    const filteredDayun = [];
    for (let i = 0; i < dayunList.length; i++) {
      const daYun = dayunList[i];
      const sYear = daYun.getStartYear();
      const eYear = daYun.getEndYear();
      const ganZhi = daYun.getGanZhi();

      // 逻辑：起止年份只要重叠了出生到80岁期间就显示，且必须有干支值
      // If DaYun period overlaps with birth to age 80 (e.g., 79-88 overlaps with birth to 80)
      if (eYear >= birthYear && sYear <= age80Year && ganZhi && ganZhi.trim() !== '') {
        filteredDayun.push({
          startYear: sYear,
          endYear: eYear,
          ganZhi: ganZhi
        });
      }
    }

    // Return object with header and dayun list
    return {
      header: result,
      dayunList: filteredDayun
    };

  } catch (error) {
    console.error('Bazi calculation error:', error);
    return `计算出错: ${error.message}`;
  }
}

File: bazi.py
#!/bin/python3

try:
    from lunar_python import Solar
except ImportError:
    print("library not found, run:")
    print("pip3 install --upgrade lunar-python --break-system-packages")
    exit(1)

from datetime import datetime

def calc_dayun(solar, gender):
    lunar = solar.getLunar()
    eight_char = lunar.getEightChar()

    is_male = gender.lower() == 'm' or gender.lower() == '1'
    yun = eight_char.getYun(is_male)

    # Generate Da Yun periods
    dayun_list = yun.getDaYun()

    # Filter dayun list to make sure we only consider valid ones with GanZhi
    valid_dayuns = [d for d in dayun_list if d.getGanZhi() and d.getGanZhi().strip()]

    if valid_dayuns:
        current_year = datetime.now().year
        active_idx = -1
        for idx, d in enumerate(valid_dayuns):
            if current_year >= d.getStartYear() and current_year <= d.getEndYear():
                active_idx = idx
                break

        # If not started yet, default to first DaYun
        if active_idx == -1 and current_year < valid_dayuns[0].getStartYear():
            active_idx = 0

        if active_idx != -1:
            # Print current DaYun
            curr = valid_dayuns[active_idx]
            print(f"{curr.getStartYear()}-{curr.getEndYear()} {curr.getGanZhi()}大运")

            # Print next DaYun if it exists
            if active_idx + 1 < len(valid_dayuns):
                nxt = valid_dayuns[active_idx + 1]
                print(f"{nxt.getStartYear()}-{nxt.getEndYear()} {nxt.getGanZhi()}大运")

# ========== Main Script ==========

if __name__ == "__main__":
    try:
        # Input section
        gender = input("Gender (m/f): ").strip().lower()
        birth_year = int(input("Birth Year: "))
        birth_month = int(input("Birth Month: "))
        birth_day = int(input("Birth Date: "))
        birth_time_input = input("Birth Time (0-23, or press Enter to skip): ").strip()

        from datetime import datetime, timedelta
        
        if birth_time_input:
            birth_hour = int(birth_time_input)
            # 处理晚子时 (23:00-23:59): 需要特殊处理
            if birth_hour == 23:
                # 晚子时：使用下一天的日期来计算农历日，小时干支用下一天的早子时
                birth_date = datetime(birth_year, birth_month, birth_day)
                next_day = birth_date + timedelta(days=1)
                solar = Solar.fromYmdHms(next_day.year, next_day.month, next_day.day, 0, 0, 0)
                # 小时干支用下一天的早子时 (hour 0 of next day)
                solar_for_hour = Solar.fromYmdHms(next_day.year, next_day.month, next_day.day, 0, 0, 0)
            else:
                # 早子时 (0) 和其他时辰：使用当前日期
                solar = Solar.fromYmdHms(birth_year, birth_month, birth_day, birth_hour, 0, 0)
                solar_for_hour = solar
        else:
            solar = Solar.fromYmd(birth_year, birth_month, birth_day)
            solar_for_hour = solar

        # Get Lunar object and GanZhi
        lunar = solar.getLunar()
        year_gz = lunar.getYearInGanZhi()
        month_gz = lunar.getMonthInGanZhi()
        day_gz = lunar.getDayInGanZhi()

        if not birth_time_input:
            print("\nHour not specified. Showing 12 possible combinations:")
            for i in range(0, 24, 2):
                temp_solar = Solar.fromYmdHms(birth_year, birth_month, birth_day, i, 0, 0)
                temp_lunar = temp_solar.getLunar()
                hour_gz = temp_lunar.getTimeInGanZhi()
                time_range = f"{i:02d}:00-{i+1:02d}:59"
                # print(f"{time_range} → {birth_year}年 {year_gz}年 {month_gz}月 {day_gz}日 {hour_gz}时")
                print(f"{birth_year}年 {year_gz}年 {month_gz}月 {day_gz}日 {hour_gz}时")

        print("\n【八字命盘 十年大运】\n")
        print(f"{birth_year} {year_gz}年 {month_gz}月 {day_gz}日", end="")

        if birth_time_input:
            # 获取小时干支（对于晚子时，使用下一天的早子时的小时干支）
            hour_lunar = solar_for_hour.getLunar()
            hour_gz = hour_lunar.getTimeInGanZhi()
            print(f" {hour_gz}时")
        else:
            print("")

        # Calculate and print Da Yun
        calc_dayun(solar, gender)

    except KeyboardInterrupt:
        print("\n [i] Cancelled")

    except ValueError:
        print("\n [i] Invalid input")
